src/protocol/graphics.ts

```diff
@@ -423,3 +416,66 @@ export async function getFirstMeaningfulPaint(limit: number = 10) {
     }
   }
 }
+
+// precache this many milliseconds
+const precacheTime = 5000;
+// startTime of the currently running precacheScreenshots() call
+let precacheStartTime = -1;
+
+export function precacheScreenshots(startTime: number): UIThunkAction {
+  return async ({ dispatch, getState }) => {
+    await paintPointsWaiter;
+
+    const recordingDuration = selectors.getRecordingDuration(getState());
+    if (!recordingDuration) {
+      return;
+    }
+
+    startTime = snapTimeForPlayback(startTime);
+    if (startTime === precacheStartTime) {
+      return;
+    }
+    if (startTime < precacheStartTime) {
+      dispatch(actions.setPlaybackPrecachedTime(startTime));
+    }
+    precacheStartTime = startTime;
+
+    const endTime = Math.min(startTime + precacheTime, recordingDuration);
+    for (let time = startTime; time < endTime; time += snapInterval) {
+      const index = mostRecentIndex(gPaintPoints, time);
+      if (index === undefined) {
+        return;
+      }
+
+      const paintHash = gPaintPoints[index].paintHash;
+      if (!screenshotCache.hasScreenshot(paintHash)) {
+        const graphicsPromise = getGraphicsAtTime(time, true);
+
+        const precachedTime = Math.max(time - snapInterval, startTime);
+        if (precachedTime > selectors.getPlaybackPrecachedTime(getState())) {
+          dispatch(actions.setPlaybackPrecachedTime(precachedTime));
+        }
+
+        await graphicsPromise;
+
+        if (precacheStartTime !== startTime) {
+          return;
+        }
+      }
+    }
+
+    let precachedTime = endTime;
+    if (mostRecentIndex(gPaintPoints, precachedTime) === gPaintPoints.length - 1) {
+      precachedTime = recordingDuration;
+    }
+    if (precachedTime > selectors.getPlaybackPrecachedTime(getState())) {
+      dispatch(actions.setPlaybackPrecachedTime(precachedTime));
+    }
+  };
+}
+
+// Snap time to 50ms intervals, snapping up.
+const snapInterval = 50;
+export function snapTimeForPlayback(time: number) {
+  return time + snapInterval - (time % snapInterval);
```

Comment:
I just realized I made the same error in my code, but this will align `0` up to `50`, and we probably want `0` to map to `0`.  I can't think if a super-clever expression for this, so my best idea is:

```
const rem = time % snapInterval;
return time + (rem > 0 ? snapInterval - rem : 0);
```